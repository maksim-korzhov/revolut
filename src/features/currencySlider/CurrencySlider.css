.slick-dots {
    bottom: 0;
}

.slick-dots li button:before {
    color: #ccc;
}

.slick-dots li.slick-active button:before {
    color: #fff;
}
