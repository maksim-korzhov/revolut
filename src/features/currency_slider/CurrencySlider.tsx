import * as React from "react";
import Slider from "react-slick";
import "slick-carousel/slick/slick.css";
import "slick-carousel/slick/slick-theme.css";
import "./currency_slider.css";

import CurrencySlide from "../currency_slide/CurrencySlide";

const settings = {
  arrows: false,
  dots: true,
  easing: "ease-in-out",
  infinite: true,
  mobileFirst: true,
  speed: 500,
  slidesToShow: 1,
  slidesToScroll: 1,
};

interface IProps {}

const CurrencySlider: React.FunctionComponent<IProps> = () => {
  return (
    <section className="exchange">
      <Slider {...settings}>
        <div className="slider-item">
          <CurrencySlide name="RUB" sign="$" value="0.01" />
        </div>
        <div className="slider-item">
          <CurrencySlide name="GBP" sign="$" value="58.01" />
        </div>
        <div className="slider-item">
          <CurrencySlide name="EUR" sign="$" value="63.01" />
        </div>
        <div className="slider-item">
          <CurrencySlide name="USD" sign="$" value="63.01" />
        </div>
      </Slider>
    </section>
  );
};

export default CurrencySlider;
