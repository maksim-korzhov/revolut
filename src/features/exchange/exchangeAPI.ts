// A mock function to mimic making an async request for data
export function fetchRates() {
  return fetch(
    `https://openexchangerates.org/api/latest.json?app_id=48a4f37d75e54e7a8498de71bb20c987`
  )
    .then((response) => {
      return response.json();
    })
    .then((response) => {
      return response.rates;
    })
    .catch((error) => {
      console.log("Err", error);
    });
}
