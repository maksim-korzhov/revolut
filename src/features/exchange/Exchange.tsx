import React from "react";
import CurrencySlider from "../currencySlider/CurrencySlider";
import CurrencyInput from "../currencyInput/CurrencyInput";
import { useAppSelector, useAppDispatch } from "../../app/hooks";
import {
  selectFromToWallets,
  selectWallets,
  selectCurrentRate,
  setWallets,
  exchangeValue,
  loadRatesAsync,
} from "./exchangeSlice";
import { getRatedValue } from "../../helpers/converters";

interface IProps {}

const Exchange: React.FunctionComponent<IProps> = () => {
  const wallets = useAppSelector(selectWallets);
  const [fromWallet, toWallet] = useAppSelector(selectFromToWallets);
  const fromSign = wallets[fromWallet].sign;
  const toSign = wallets[toWallet].sign;
  const currentRate = useAppSelector(selectCurrentRate);
  const dispatch = useAppDispatch();

  const [fromValue, setFromValue] = React.useState(0);
  const [toValue, setToValue] = React.useState(0);

  /**
   * Calculate "to" value using the current rate
   * @param value
   */
  const calculateValue = (value: number) => {
    setFromValue(value);

    if (value) {
      const newToValue = getRatedValue(currentRate, value);
      setToValue(newToValue);
    } else {
      setToValue(value);
    }
  };

  /** Load base currency rates */
  React.useEffect(() => {
    dispatch(loadRatesAsync());

    const interval = window.setInterval(
      () => dispatch(loadRatesAsync()),
      10000
    );

    return () => window.clearInterval(interval);
  }, []);

  /** Reset values on from wallet changes */
  React.useEffect(() => {
    setFromValue(0);
    setToValue(0);
  }, [fromWallet]);

  // /** Calculate values on every wallet changes */
  React.useEffect(() => {
    calculateValue(fromValue);
  }, [toWallet, calculateValue]);

  return (
    <>
      <header className="header">
        <div className="current-rate">
          {`${fromSign}1 = ${toSign}${currentRate.toFixed(4)}`}
        </div>
      </header>

      <div className="slider-container">
        <CurrencySlider
          wallets={wallets}
          current={fromWallet}
          onChange={(from: string) => dispatch(setWallets([from, toWallet]))}
        />

        <CurrencyInput
          maxValue={wallets[fromWallet].total}
          name="from"
          prefix="â€“"
          onChange={calculateValue}
          value={fromValue}
        />
      </div>

      <div className="slider-container">
        <CurrencySlider
          wallets={wallets}
          current={toWallet}
          onChange={(to: string) => dispatch(setWallets([fromWallet, to]))}
        />

        <CurrencyInput disabled name="to" prefix="+" value={toValue} />
      </div>

      <footer className="footer">
        <button
          className="btn btn-exchange"
          disabled={!fromValue || fromWallet === toWallet}
          onClick={() => {
            dispatch(exchangeValue(fromValue));
            setFromValue(0);
            setToValue(0);
          }}
        >
          Exchange
        </button>
      </footer>
    </>
  );
};

export default Exchange;
