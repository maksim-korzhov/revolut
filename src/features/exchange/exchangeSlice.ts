import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "../../app/store";
import { fetchRates } from "./exchangeAPI";
import currency from "currency.js";
import { getRatedValue } from "../../helpers/converters";

export enum LoadingState {
  IDLE,
  LOADING,
  FAILED,
}

export enum RatesEnum {
  USD = "USD",
  EUR = "EUR",
  RUB = "RUB",
  GBP = "GBP",
}

export interface CurrencyData {
  name: RatesEnum;
  sign: string;
  total: number;
}

export type Wallets = Record<RatesEnum, CurrencyData>;
export interface ExchangeState {
  wallets: Wallets;
  rates: {
    from: RatesEnum;
    to: RatesEnum;
    data: Record<RatesEnum, number>;
    status: LoadingState;
  };
}
export type CurrencyShort = Record<
  RatesEnum,
  Pick<CurrencyData, "name" | "sign" | "total">
>;
export type ExchangeStateShort = Record<RatesEnum, CurrencyShort>;

export const initialState: ExchangeState = {
  wallets: {
    USD: {
      name: RatesEnum.USD,
      sign: "$",
      total: 45.43,
    },
    EUR: {
      name: RatesEnum.EUR,
      sign: "€",
      total: 132.44,
    },
    RUB: {
      name: RatesEnum.RUB,
      sign: "₽",
      total: 10007.43,
    },
    GBP: {
      name: RatesEnum.GBP,
      sign: "£",
      total: 21.1,
    },
  },
  rates: {
    from: RatesEnum.USD,
    to: RatesEnum.USD,
    status: LoadingState.IDLE,
    data: {
      EUR: 0.82,
      RUB: 73.6975,
      GBP: 0.71,
      USD: 1,
    },
  },
};

// The function below is called a thunk and allows us to perform async logic. It
// can be dispatched like a regular action: `dispatch(incrementAsync(10))`. This
// will call the thunk with the `dispatch` function as the first argument. Async
// code can then be executed and other actions can be dispatched. Thunks are
// typically used to make async requests.
export const loadRatesAsync = createAsyncThunk(
  "counter/loadRates",
  async () => {
    const response = await fetchRates();

    if (!response) throw new Error("Can not load rates");
    // The value we return becomes the `fulfilled` action payload
    // return response.data;
    return response;
  }
);

export const exchangeSlice = createSlice({
  name: "exchange",
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {
    exchangeValue: (state, action: PayloadAction<number>) => {
      const [fromWallet, toWallet] = selectFromToWallets({ exchange: state });
      const rate = selectCurrentRate({ exchange: state });

      // Get money from the current "from" wallet
      state.wallets[fromWallet as RatesEnum].total = currency(
        state.wallets[fromWallet as RatesEnum].total
      ).subtract(action.payload).value;

      // Move money to the current from wallet
      state.wallets[toWallet as RatesEnum].total = currency(
        state.wallets[toWallet as RatesEnum].total
      ).add(getRatedValue(rate, action.payload)).value;
    },
    setWallets: (state, action: PayloadAction<[string, string]>) => {
      state.rates.from = action.payload[0] as RatesEnum;
      state.rates.to = action.payload[1] as RatesEnum;
    },
  },
  // The `extraReducers` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers: (builder) => {
    builder
      .addCase(loadRatesAsync.pending, (state) => {
        state.rates.status = LoadingState.LOADING;
      })
      .addCase(loadRatesAsync.fulfilled, (state, action) => {
        state.rates.status = LoadingState.IDLE;
        state.rates.data = Object.keys(state.wallets).reduce(
          (acc: any, key: string) => {
            if (action.payload[key]) {
              // Fake number to emulate different rates
              const randomMultiplier =
                key === RatesEnum.USD ? 1 : Math.random() * 0.1 + 1;

              acc[key] = action.payload[key] * randomMultiplier;
            }

            return acc;
          },
          {}
        );
      })
      .addCase(loadRatesAsync.rejected, (state, action) => {
        state.rates.status = LoadingState.FAILED;
      });
  },
});

export const { exchangeValue, setWallets } = exchangeSlice.actions;

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state: RootState) => state.counter.value)`
export const selectFromToWallets = (state: RootState) => {
  const rates = state.exchange.rates;
  return [rates.from, rates.to];
};

// Select base wallets data for the slider
export const selectWallets = ({ exchange }: RootState) => {
  return Object.keys(exchange.wallets).reduce((acc: any, key: string) => {
    const currentWallet = exchange.wallets[key as RatesEnum];
    acc[key as RatesEnum] = {
      name: currentWallet.name,
      sign: currentWallet.sign,
      total: currentWallet.total,
    };

    return acc;
  }, {} as any);
};

/**
 * Function receives convert currency rate from rateFromName currency
 * to the rateToName currency
 * @param rateFromNames
 * @param rateToName
 * @returns
 */
export const selectCurrentRate = ({ exchange }: RootState) => {
  const rateFrom = exchange.rates.from;
  const rateTo = exchange.rates.to;

  if (rateFrom === rateTo) return 1;

  const from = exchange.rates.data[rateFrom as RatesEnum];
  const to = exchange.rates.data[rateTo as RatesEnum];

  return to / from;
};

/** Return loading rate status */
export const selectStatus = ({ exchange }: RootState) => {
  return exchange.rates.status;
};

export default exchangeSlice.reducer;
