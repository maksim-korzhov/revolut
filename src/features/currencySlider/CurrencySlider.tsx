import * as React from "react";
import Slider from "react-slick";
import CurrencySlide from "../currencySlide/CurrencySlide";

import "slick-carousel/slick/slick.css";
import "slick-carousel/slick/slick-theme.css";
import "./CurrencySlider.css";

import { RatesEnum, Wallets } from "../exchange/exchangeSlice";

const settings = {
  arrows: false,
  dots: true,
  easing: "ease-in-out",
  infinite: true,
  mobileFirst: true,
  speed: 500,
  slidesToShow: 1,
  slidesToScroll: 1,
};

interface IProps {
  current: string;
  wallets: Wallets;
  onChange: Function;
}

/**
 * Helper function. Renders wallets data.
 * @param wallet
 * @returns
 */
const renderCurrencySlide = ({ name, sign, total }: any) => (
  <div className="slider-item" key={name}>
    <CurrencySlide name={name} sign={sign} value={total} />
  </div>
);

const afterChange = (
  wallets: Wallets,
  slideNumber: number,
  callback: Function
) => {
  const currentWalletKey = Object.keys(wallets)[slideNumber];
  const currentWallet = wallets[currentWalletKey as RatesEnum];
  callback(currentWallet.name);
};

/**
 * Renders slides with wallets data.
 * @param data
 * @returns
 */
const CurrencySlider: React.FunctionComponent<IProps> = ({
  wallets,
  onChange,
}) => {
  return (
    <section className="exchange">
      <Slider
        {...settings}
        afterChange={(slideNumber: number) =>
          afterChange(wallets, slideNumber, onChange)
        }
      >
        {Object.keys(wallets).map((key: string) =>
          renderCurrencySlide(wallets[key as RatesEnum])
        )}
      </Slider>
    </section>
  );
};

export default CurrencySlider;
