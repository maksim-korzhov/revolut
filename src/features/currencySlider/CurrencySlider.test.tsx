import React from "react";
import { render } from "@testing-library/react";
import CurrencySlider from "./CurrencySlider";
import { initialState } from "../exchange/exchangeSlice";

const defaultValues = {
  current: "USD",
  wallets: initialState.wallets,
  onChange: () => false,
};

describe("CurrencySlider component", () => {
  it("renders correctly with the default props", () => {
    const component = render(<CurrencySlider {...defaultValues} />);

    expect(component).toMatchSnapshot();
  });
});
