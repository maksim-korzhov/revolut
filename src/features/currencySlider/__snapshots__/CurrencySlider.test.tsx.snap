// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySlider component renders correctly with the default props 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="exchange"
      >
        <div
          class="slick-slider slick-initialized"
          dir="ltr"
        >
          <div
            class="slick-list"
          >
            <div
              class="slick-track"
              style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
            >
              <div
                aria-hidden="true"
                class="slick-slide slick-cloned"
                data-index="-1"
                style="width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    class="slider-item"
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  >
                    <div
                      class="exchange-wrapper"
                    >
                      <div
                        class="exchange-content"
                      >
                        <h3
                          class="exchange-currency"
                        >
                          GBP
                        </h3>
                        <p
                          class="exchange-inwallet"
                        >
                          You have £21.1
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="false"
                class="slick-slide slick-active slick-current"
                data-index="0"
                style="outline: none; width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    class="slider-item"
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  >
                    <div
                      class="exchange-wrapper"
                    >
                      <div
                        class="exchange-content"
                      >
                        <h3
                          class="exchange-currency"
                        >
                          USD
                        </h3>
                        <p
                          class="exchange-inwallet"
                        >
                          You have $45.43
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="true"
                class="slick-slide"
                data-index="1"
                style="outline: none; width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    class="slider-item"
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  >
                    <div
                      class="exchange-wrapper"
                    >
                      <div
                        class="exchange-content"
                      >
                        <h3
                          class="exchange-currency"
                        >
                          EUR
                        </h3>
                        <p
                          class="exchange-inwallet"
                        >
                          You have €132.44
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="true"
                class="slick-slide"
                data-index="2"
                style="outline: none; width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    class="slider-item"
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  >
                    <div
                      class="exchange-wrapper"
                    >
                      <div
                        class="exchange-content"
                      >
                        <h3
                          class="exchange-currency"
                        >
                          RUB
                        </h3>
                        <p
                          class="exchange-inwallet"
                        >
                          You have ₽10007.43
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="true"
                class="slick-slide"
                data-index="3"
                style="outline: none; width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    class="slider-item"
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  >
                    <div
                      class="exchange-wrapper"
                    >
                      <div
                        class="exchange-content"
                      >
                        <h3
                          class="exchange-currency"
                        >
                          GBP
                        </h3>
                        <p
                          class="exchange-inwallet"
                        >
                          You have £21.1
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="true"
                class="slick-slide slick-cloned"
                data-index="4"
                style="width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    class="slider-item"
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  >
                    <div
                      class="exchange-wrapper"
                    >
                      <div
                        class="exchange-content"
                      >
                        <h3
                          class="exchange-currency"
                        >
                          USD
                        </h3>
                        <p
                          class="exchange-inwallet"
                        >
                          You have $45.43
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="true"
                class="slick-slide slick-cloned"
                data-index="5"
                style="width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    class="slider-item"
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  >
                    <div
                      class="exchange-wrapper"
                    >
                      <div
                        class="exchange-content"
                      >
                        <h3
                          class="exchange-currency"
                        >
                          EUR
                        </h3>
                        <p
                          class="exchange-inwallet"
                        >
                          You have €132.44
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="true"
                class="slick-slide slick-cloned"
                data-index="6"
                style="width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    class="slider-item"
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  >
                    <div
                      class="exchange-wrapper"
                    >
                      <div
                        class="exchange-content"
                      >
                        <h3
                          class="exchange-currency"
                        >
                          RUB
                        </h3>
                        <p
                          class="exchange-inwallet"
                        >
                          You have ₽10007.43
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="true"
                class="slick-slide slick-cloned"
                data-index="7"
                style="width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    class="slider-item"
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  >
                    <div
                      class="exchange-wrapper"
                    >
                      <div
                        class="exchange-content"
                      >
                        <h3
                          class="exchange-currency"
                        >
                          GBP
                        </h3>
                        <p
                          class="exchange-inwallet"
                        >
                          You have £21.1
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <ul
            class="slick-dots"
            style="display: block;"
          >
            <li
              class="slick-active"
            >
              <button>
                1
              </button>
            </li>
            <li
              class=""
            >
              <button>
                2
              </button>
            </li>
            <li
              class=""
            >
              <button>
                3
              </button>
            </li>
            <li
              class=""
            >
              <button>
                4
              </button>
            </li>
          </ul>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="exchange"
    >
      <div
        class="slick-slider slick-initialized"
        dir="ltr"
      >
        <div
          class="slick-list"
        >
          <div
            class="slick-track"
            style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
          >
            <div
              aria-hidden="true"
              class="slick-slide slick-cloned"
              data-index="-1"
              style="width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  class="slider-item"
                  style="width: 100%; display: inline-block;"
                  tabindex="-1"
                >
                  <div
                    class="exchange-wrapper"
                  >
                    <div
                      class="exchange-content"
                    >
                      <h3
                        class="exchange-currency"
                      >
                        GBP
                      </h3>
                      <p
                        class="exchange-inwallet"
                      >
                        You have £21.1
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-hidden="false"
              class="slick-slide slick-active slick-current"
              data-index="0"
              style="outline: none; width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  class="slider-item"
                  style="width: 100%; display: inline-block;"
                  tabindex="-1"
                >
                  <div
                    class="exchange-wrapper"
                  >
                    <div
                      class="exchange-content"
                    >
                      <h3
                        class="exchange-currency"
                      >
                        USD
                      </h3>
                      <p
                        class="exchange-inwallet"
                      >
                        You have $45.43
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-hidden="true"
              class="slick-slide"
              data-index="1"
              style="outline: none; width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  class="slider-item"
                  style="width: 100%; display: inline-block;"
                  tabindex="-1"
                >
                  <div
                    class="exchange-wrapper"
                  >
                    <div
                      class="exchange-content"
                    >
                      <h3
                        class="exchange-currency"
                      >
                        EUR
                      </h3>
                      <p
                        class="exchange-inwallet"
                      >
                        You have €132.44
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-hidden="true"
              class="slick-slide"
              data-index="2"
              style="outline: none; width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  class="slider-item"
                  style="width: 100%; display: inline-block;"
                  tabindex="-1"
                >
                  <div
                    class="exchange-wrapper"
                  >
                    <div
                      class="exchange-content"
                    >
                      <h3
                        class="exchange-currency"
                      >
                        RUB
                      </h3>
                      <p
                        class="exchange-inwallet"
                      >
                        You have ₽10007.43
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-hidden="true"
              class="slick-slide"
              data-index="3"
              style="outline: none; width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  class="slider-item"
                  style="width: 100%; display: inline-block;"
                  tabindex="-1"
                >
                  <div
                    class="exchange-wrapper"
                  >
                    <div
                      class="exchange-content"
                    >
                      <h3
                        class="exchange-currency"
                      >
                        GBP
                      </h3>
                      <p
                        class="exchange-inwallet"
                      >
                        You have £21.1
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-hidden="true"
              class="slick-slide slick-cloned"
              data-index="4"
              style="width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  class="slider-item"
                  style="width: 100%; display: inline-block;"
                  tabindex="-1"
                >
                  <div
                    class="exchange-wrapper"
                  >
                    <div
                      class="exchange-content"
                    >
                      <h3
                        class="exchange-currency"
                      >
                        USD
                      </h3>
                      <p
                        class="exchange-inwallet"
                      >
                        You have $45.43
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-hidden="true"
              class="slick-slide slick-cloned"
              data-index="5"
              style="width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  class="slider-item"
                  style="width: 100%; display: inline-block;"
                  tabindex="-1"
                >
                  <div
                    class="exchange-wrapper"
                  >
                    <div
                      class="exchange-content"
                    >
                      <h3
                        class="exchange-currency"
                      >
                        EUR
                      </h3>
                      <p
                        class="exchange-inwallet"
                      >
                        You have €132.44
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-hidden="true"
              class="slick-slide slick-cloned"
              data-index="6"
              style="width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  class="slider-item"
                  style="width: 100%; display: inline-block;"
                  tabindex="-1"
                >
                  <div
                    class="exchange-wrapper"
                  >
                    <div
                      class="exchange-content"
                    >
                      <h3
                        class="exchange-currency"
                      >
                        RUB
                      </h3>
                      <p
                        class="exchange-inwallet"
                      >
                        You have ₽10007.43
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-hidden="true"
              class="slick-slide slick-cloned"
              data-index="7"
              style="width: 0px;"
              tabindex="-1"
            >
              <div>
                <div
                  class="slider-item"
                  style="width: 100%; display: inline-block;"
                  tabindex="-1"
                >
                  <div
                    class="exchange-wrapper"
                  >
                    <div
                      class="exchange-content"
                    >
                      <h3
                        class="exchange-currency"
                      >
                        GBP
                      </h3>
                      <p
                        class="exchange-inwallet"
                      >
                        You have £21.1
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <ul
          class="slick-dots"
          style="display: block;"
        >
          <li
            class="slick-active"
          >
            <button>
              1
            </button>
          </li>
          <li
            class=""
          >
            <button>
              2
            </button>
          </li>
          <li
            class=""
          >
            <button>
              3
            </button>
          </li>
          <li
            class=""
          >
            <button>
              4
            </button>
          </li>
        </ul>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
